SoundFile: fix SynthDef and Helpfile to match

patrickdupuis opened the pull request:
I initially wanted to update SoundFile's `cue` method documentation to correctly indicate the use of VDiskIn inside it's SynthDef. What I ended up doing is rewriting the SynthDef in the class file and in the documentation. The major difference is the removal of the redundant `Linen`. I couldn't understand why there needed to be two envelopes (one sustaining with gate, the other fixed duration without gate, both set to `doneAction:2`). Maybe I'm misunderstanding...? The rest of the changes are mostly cosmetic.

patrickdupuis commented:
It's possible to send a custom event to .cue that sets these arguments to some other values. I'm afraid that if we change the argument names someone's code will break. It's highly unlikely, though... We can introduce breaking changes in 3.10. Maybe it's ok here?

jamshark70 commented:
One gotcha is that `Linen.kr` is gated, while `Env.linen` is not. So these changes eliminate the possibility of ending the synth by setting `gate` to 0.

Technically, it was never really correct to pass `sustain` as an input. In events, `sustain` is in beats while the server knows only seconds. But I expect no way to fix that without breaking code.

patrickdupuis commented:
Oh! I had assumed that I could simply move the `gate` argument to the `Env.linen` and get rid of `Linen` but I see now that it doesn't work. I'll bring back the gated envelope. Thanks for pointing that out @jamshark70.

patrickdupuis commented:
I think the SynthDef is clearer now even with the two envelopes. Since `sustain` was a mistake, I've changed that argument's name to `sustainTime` (matches Env.linen's arg name), and changed `ar` and `dr` to `atk` and `rel`.

P.S. I remembered that there's a way to force release envelopes by setting gate below `-1`. I couldn't find an elegant way of using this to avoid bringing back the `Linen`. Makes me think we need some sort of UGen to help make better use of this feature.

jamshark70 commented:
The one other thing I might suggest is to put an attack time of 0 in one or the other envelope. Two envelopes with the same linear attack, multiplied, produces an x.squared curve... Probably a concern mainly for accuracy of documentation (it would be weird to have to explain why the attack is curved while the release is linear).

Minor point, maybe ok to ignore it.

telephon commented:
I don't see why the early gate should lead to a different release than the regular one. It would be better to multiply the gate by something like `Trig1.kr(Impulse.ar(0), sustainTime)`.


> I remembered that there's a way to force release envelopes by setting gate below -1.

Yes, you can determine the release time by the absolute of the value you set it to (e.g. -2.5 is 2.5 sec).

patrickdupuis commented:
@jamshark70 good point about the attack ramp. I forgot to mention that in my last commit, I changed the `atk` and `rel` argument default values to `0`. I didn't realise that I had to hardcode `Linen`'s attack to `0` and make use of `atk` in the `Env.linen` only.

patrickdupuis commented:
@telephon I don't understand what you mean by the different releases. Aren't the two envelopes set to have the same release times? They both make use of the `rel` argument.

telephon commented:
When you write:

```supercollider
f = SoundFile(Platform.resourceDir +/+ "sounds/a11wlk01.wav")
e = f.cue;
e.play;
e.release(10);
```

The gate `-10` is sent, but overridden by the `rel` argument. This is not a huge thing, but still I find the two envelopes not a very good choice.

I'd prefer something like:
```supercollider
(
SynthDef(\diskIn2, { |out, amp=1, bufnum, sustainTime, atk=0, rel=0, gate=1|
	var sig = VDiskIn.ar(2, bufnum, BufRateScale.kr(bufnum));
	var env, earlyGate;
	earlyGate = EnvGen.kr(Env([1, 1, 0], [(sustainTime-atk-rel).max(0), 0]);
	env = EnvGen.kr(Env.asr(atk, amp, rel), gate * earlyGate, doneAction: Done.freeSelf);
	Out.ar(out, sig * env)
}).add;
)
```

(untested)

But if you have no time to fix this, no problem!

jamshark70 commented:
> The gate -10 is sent, but overridden by the rel argument.

It seems more likely to me that Linen simply doesn't implement the negative gate early release. By definition, the early release should override the envelope times.

Maybe Linen should...?

telephon commented:
The normal envelope in this SynthDef doesn't gave a `gate`  argument, so it doesn't respond to that anyway.

patrickdupuis commented:
@jamshark70 Linen can be force released with a negative gate just fine over here. What exactly did you try that wasn't working?

The best solution is the one @telephon suggested. Using a fixed duration envelope as an envelope for the gate of the sustaining envelope seems to work correcty. I've pushed this change to the PR.

jamshark70 commented:
> Linen can be force released with a negative gate just fine over here. What exactly did you try that wasn't working?

I didn't. I was responding to Julian's comment that the `rel` argument "overrides" the release time calculated from the negative gate.

If Linen supports negative gate in the same way that EnvGen does, then `rel` can't override it.

So I don't know which one of you is correct now.

To be honest, I'm soul-crushingly busy this week, so I'm dropping out of this discussion... If there's still disagreement, draw up a spec before messing around with code.

patrickdupuis commented:
No worries @jamshark70. @telephon please give this another review when you get the chance.
